### ISSUE TITLE
Mismatched RSA Key Thumbprints

### ISSUE THREAD

**Issue by MichaelStankiewicz** [chef/supermarket#1688]
> For some context, I have OAuth setup between my Chef Server and Supermarket server so that authorization to the supermarket instance is done via Chef Server.

I'm getting an authorization error when attempting to use knife to upload cookbooks to my internal supermarket server (this used to work before).  When I log onto the web GUI of the supermarket portal and go to Manage Profiles | Keys, I see information outlining a procedure for validating the Chef Server key thumbprint against the thumbprint hash that the Chef Supermarket has for me.  When I execute the openssl command as presented, I clearly can see that the thumbprints **do not match**.  

I've found blogs online of people with similar issues that state you can logon/logoff of the supermarket web console to initiate a refresh of the keys from the Chef Server.  But I've done this several times to no avail; I've also even rotated my Chef Server user key but none of this has fixed the issue.

**Comment by robbkidd** [comment]
Are you using multiple keys with your account on the Chef Server?

**Comment by MichaelStankiewicz** [comment]
No, but I have an update.  I noticed that the date/time were skewed on my Chef Server (it turned out that the time was a couple days back from the real date/time).  After rereading the Chef documentation for private supermarket, I noticed that it was a prerequisite to have the clocks sync'd between the two services (this is in a dev lab so no strategic clock syncing going on).  After I made this change, I can now use knife successfully to upload cookbooks.  

Even though I can now use knife to upload cookbooks again, the thumbprint mismatch is still observed in the supermarket console.  Unless told otherwise, I'd like to leave this issue open as it is still not reporting properly.

**Comment by robbkidd** [comment]
We'll need more information to determine why there is a mismatch. If you're able to upload to your supermarket with knife, your keys match. The remaining fingerprint issue must be something to do with the openssl command invocations.

```
openssl rsa -in your/private_key.pem -pubout -outform DER | openssl md5 -c
```

In the example command, `your/private_key.pem` is expected to be a PEM file containing the private key associated with the public key that is on the Chef Server and in Supermarket. To troubleshoot, run the portion of the command prior to the `|` and check the exit status of the openssl command to confirm it was able to parse the key given.

Some examples. Note: in the two fail-to-parse examples that the "fingerprints" generated for the nothingness sent into the command pipe match each other. 

#### a non existent file

```
> openssl rsa -in your/private_key.pem -pubout -outform DER
Error opening Private Key your/private_key.pem
96710:error:02001002:system library:fopen:No such file or directory:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-64.50.6/src/crypto/bio/bss_file.c:356:fopen('your/private_key.pem','r')
96710:error:20074002:BIO routines:FILE_CTRL:system lib:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-64.50.6/src/crypto/bio/bss_file.c:358:
unable to load Private Key
> echo $?
1

> openssl rsa -in your/private_key.pem -pubout -outform DER | openssl md5 -c
d4:1d:8c:d9:8f:00:b2:04:e9:80:09:98:ec:f8:42:7e
```

#### a public key 

```
> openssl rsa -in ~/.chef/the_public_half_only.pub -pubout -outform DER
unable to load Private Key
96903:error:0906D06C:PEM routines:PEM_read_bio:no start line:/BuildRoot/Library/Caches/com.apple.xbs/Sources/OpenSSL098/OpenSSL098-64.50.6/src/crypto/pem/pem_lib.c:648:Expecting: ANY PRIVATE KEY
> echo $?
1

> openssl rsa -in ~/.chef/the_public_half_only.pub -pubout -outform DER | openssl md5 -c
d4:1d:8c:d9:8f:00:b2:04:e9:80:09:98:ec:f8:42:7e
```

#### a real key

```
> openssl rsa -in ~/.chef/a_real_key.pem -pubout -outform DER
writing RSA key
{gobbledyguk}%
> echo $?
0
```

**Comment by MichaelStankiewicz** [comment]
Interesting.  So I ran this from the cmd.exe shell (I'm running on Windows), and the key came back just fine.  However when I ran this from PowerShell (where I've been running it from up until now), this was resulting in the mismatch.

Therefore, there seems to be some parsing issue within PowerShell that is making this convert incorrectly.  I'm not sure why, but at least we know it's not a problem with the supermarket console but rather the method by which I was attempting to derive the thumbprint on my workstation.  

I think this one can be closed.  I appreciate your help with this one.

**Comment by robbkidd** [comment]
You're very welcome!

That's great to know about the PowerShell context. I think it's fair to keep this issue open and reword it around sussing out what's different about running the commands as currently given in a PowerShell session and update the instructions if we find better way.
